ruby/lib/clone_handler.rb

```diff
@@ -0,0 +1,122 @@
+require 'json'
+require 'net/http'
+
+class CloneHandler
+  PER_PAGE = 30.freeze
+  BASE_URL = 'https://api.github.com'
+
+  def self.get_username
+    Kernel.puts 'Please enter your Github username'
+    STDOUT.flush
+    STDIN.gets.chomp
+  end
+
+  def initialize(username)
+    @username = username
+    @errors = []
+  end
+
+  def clone
+    validate_clonable
+    if errors?
+      print_errors
+      return false
+    end
+
+    Kernel.system(final_command)
+  end
+
+  private
+
+  attr_reader :username, :errors
+
+  def final_command
+    @final_command ||= clone_commands.flatten.join(' & ')
+  end
+
+  def clone_commands
+    repo_info.map { |info| build_repo_clone_command(info) }
+  end
+
+  def repo_info
+    @repo_info ||= page_count.times.map { |index| repo_info_for_page(index + 1) }.flatten.compact
+  end
+
+  def build_repo_clone_command(info)
+    return if info['clone_url'].nil?
+
+    "git clone #{info['clone_url']} #{username}/#{info['name']}"
+  end
+
+  def repo_info_for_page(page)
+    response = Net::HTTP.get_response(URI("#{user_info_url}/repos?page=#{page}"))
+
+    JSON.parse(response.body)
+  rescue SocketError => _e
+    errors.push('Network error')
+  end
+
+  def page_count
+    (user_info&.dig('public_repos').to_i / PER_PAGE.to_f).ceil
+  end
+
+  def user_info
+    @user_info ||= begin
+      response = Net::HTTP.get_response(URI(user_info_url))
+
+      JSON.parse(response.body)
+    rescue SocketError => _e
+      @user_info = {'message' => 'Network error'}
+    end
+  end
+
+  def user_info_url
+    @user_info_url ||= "#{BASE_URL}/users/#{username}"
+  end
+
+  def errors?
+    errors.length > 0
+  end
+
+  def validate_clonable
+    validate_username
+    validate_user_info
+    validate_repo_info
+  end
+
+  def validate_username
+    return if !username.nil? && username.length > 0
+
+    errors.push('No username provided')
+  end
+
+  def validate_user_info
+    return if errors?
+
+    error_message = user_info&.dig('message')
+
+    if error_message == 'Not Found'
+      errors.push('Account does not exist')
+    elsif error_message&.include?('API rate limit exceeded')
+      errors.push('API rate limit exceeded')
+    elsif error_message == 'Network error'
+      errors.push('Network error')
+    elsif user_info&.dig('public_repos').to_i == 0
+      errors.push('No repositories found at this account')
+    end
+  end
+
+  def validate_repo_info
+    return if errors?
+
+    repo_info.each do |info|
+      if info&.dig('message')&.include?('API rate limit exceeded')
+        errors.push('API rate limit exceeded')
+      end
```

Comment:
I think we do not need this if statement here.

Reply:
This is iterating through all the repo info responses. Some may have succeeded. I don't want to add an error for successes. Can you elaborate?